### ISSUE TITLE
ConfigurePrimaryHttpMessageHandler<T>() not loading handler in ASP.NET Core 2.2

### ISSUE THREAD

**Issue by bkaidbb** [dotnet/extensions#851]
### Describe the bug
Registering an HttpMessageHandler by generic type parameter is causing that handler to not be loaded.  

### To Reproduce
1. Using version '2.2' of package 'Microsoft.AspNetCore.App', and 'Microsoft.Extensions.Http'
2. Run this code at startup:
```
services.AddTransient<MyClientHttpMessageHandler>();
services
    .AddHttpClient<IMyClient, MyClient>()
    .ConfigurePrimaryHttpMessageHandler<MyClientHttpMessageHandler>();
```
3. Make an http client 
4. Error is that MyClientHttpMessageHandler won't be loaded or invoked.

### Expected behavior
The handler is properly invoked when the client is used.

### Workaround
```
services
    .AddHttpClient<IMyClient, MyClient>()
    .ConfigurePrimaryHttpMessageHandler(s => s.GetRequiredService<MyClientHttpMessageHandler>())
```

### Additional context
This was the only issue I ran into upgrading a large project from netcoreapp2.1 to netcoreapp2.2. 

.NET Core SDK (reflecting any global.json):
 Version:   2.2.101
 Commit:    236713b0b7

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.101\

Host (useful for support):
  Version: 2.2.0
  Commit:  1249f08fed

**Comment by satano** [comment]
We hit this issue as well. But we found the other side of its behaviour. Our message handler is used in default (not named) `HttpClient`, where it should not be used.

The problem is, that `ConfigurePrimaryHttpMessageHandler<THandler>()` is not using builder name as the other two `ConfigurePrimaryHttpMessageHandler()` methods. So effectively, it registers this handler not for the named client, but for default one.

- [ConfigurePrimaryHttpMessageHandler(builder, configureHandler)](https://github.com/aspnet/Extensions/blob/master/src/HttpClientFactory/Http/src/DependencyInjection/HttpClientBuilderExtensions.cs#L228) - using `builder.Name`
- [ConfigurePrimaryHttpMessageHandler\<THandler\>(builder)](https://github.com/aspnet/Extensions/blob/master/src/HttpClientFactory/Http/src/DependencyInjection/HttpClientBuilderExtensions.cs#L259) - not using `builder.Name`

I can preprare a PR if you want.

**Comment by khellang** [comment]
To be honest, I think the generic overload should just call the non-generic overload:

```csharp
public static IHttpClientBuilder ConfigurePrimaryHttpMessageHandler<THandler>(this IHttpClientBuilder builder)
    where THandler : HttpMessageHandler
{
    return builder.ConfigurePrimaryHttpMessageHandler(p => p.GetRequiredService<THandler>());
}
```

**Comment by satano** [comment]
@khellang None of the overload let you configure the message handler. The argument in the other two overloads is named `configureHandler` but it is a factory function for creating (and configuring) a message handler, not for configuring already created one. In the generic `<THandler>` overload, the `THandler` type is message handler itself, so it is assumed the instance will be configured once created.

**Comment by khellang** [comment]
> None of the overload let you configure the message handler. [...] it is a factory function for creating (and configuring) a message handler, not for configuring already created one.

So which one is it? It can't be both 😜

The other overloads lets you create and configure a handler (after its constructor has been called). The generic overload doesn't. Simple as that 😂

**Comment by satano** [comment]
Well. In the generic overload the handler is obtained from service container, so the container should return it fully configured. In the simples form, it will be _self configured_ in its constructor. If you really need some more configuration, just register in the container the factory method which will create and configure the handler. For me, it really makes sense - making an overload for it would be just redundant. 

What you propose is something like this: 

``` csharp
// In reality, configureHandler would have another argument for service provider.
ConfigurePrimaryHttpMessageHandler<THandler>(this IHttpClientBuilder builder, Action<THandler> configureHandler) where THandler : HttpMessageHandler
{
    ...
    var handler = services.GetRequiredService<THandler>();
    configureHandler(handler);
    ...
}

// With using:
services.AddTransient<CustomMessageHandler>();

.ConfigurePrimaryHttpMessageHandler<CustomMessaheHandler>(handler =>
{
    // Do handler configuration.
});

// But it is the same as this, which you can do now.
services.AddTransient<CustomMessageHandler>(serviceProvider =>
{
    var handler = new CustomMessageHandler();
    // Do handler configuration.
    return handler;
});

.ConfigurePrimaryHttpMessageHandler<CustomMessaheHandler>();
```

Anyway. This discussion is not the point of this issue. 😉

**Comment by jonathann92** [comment]
Hi guys I ran into this issue today while upgrading to dotnet core 2.2. I am using netcoreapp2.2 and `<PackageReference Include="Microsoft.Extensions.Http" Version="2.2.0" />`. 

I saw that you guys have a new release a couple weeks ago: https://github.com/aspnet/Extensions/releases/tag/v2.2.2

How long will it take to be available in nuget?

**Comment by StevenLiekens** [comment]
Until the fix is published to NuGet, the code from https://github.com/aspnet/Extensions/issues/851#issuecomment-456119041 can be used as a workaround.

Example:

```csharp
services.AddTransient<MyClientHttpMessageHandler>();
services
    .AddHttpClient<IMyClient, MyClient>()
    .ConfigurePrimaryHttpMessageHandler(sp => sp.GetService<MyClientHttpMessageHandler>());
```

**Comment by cmmizer** [comment]
We ran into this on a patch version update of Microsoft.AspnetCore.All (2.1.5 to 2.1.9). 

![image](https://user-images.githubusercontent.com/42183463/61396682-3b99b280-a896-11e9-817c-81e0104ce41b.png)

**Comment by matt-lethargic** [comment]
Pardon my potential ignorance on this subject, but I thought the entire reason why we've got HttpClientFactory in .Net Core is help stop people newing up lots of HttpClient instances and using up resources. And I thought the issue really came from having many handlers and not the client itself. 

If the work around for this is to use AddTransient then aren't we creating this same problem and shouldn't it be AddSingleton?

**Comment by totht91** [comment]
@matt-lethargic No, don't register as singleton because you will get `ObjectDisposedException` after the its lifetime has expired. `AddTransient` is the solution because the `IHttpClientFactory` only requests you instance with `sp => sp.GetRequiredService<MyClientHttpMessageHandler>()` if handlers lifetime has been expired.

**Comment by totht91** [comment]
@rynowak Why is this issue still open? I can see that this issue is fixed in https://github.com/aspnet/Extensions/pull/988 but the Microsoft.Extensions.Http v2.2.2 package has not been released. Is this released in 3.0? (no milestone assigned to this issue)

**Comment by anurse** [comment]
This does appear to be fixed in 3.0 (updating milestone).
